File: examples/nostr-basics/01-event-envelope/go/main.go
package main

import (
	"encoding/json"
	"fmt"

	nostr "github.com/nbd-wtf/go-nostr"
)

func main() {
	secretKey := nostr.GeneratePrivateKey()
	publicKey, err := nostr.GetPublicKey(secretKey)
	if err != nil {
		panic(err)
	}

	event := nostr.Event{
		PubKey:    publicKey,
		CreatedAt: nostr.Now(),
		Kind:      1,
		Tags:      nostr.Tags{{"t", "nip-01"}},
		Content:   "hello from the NIP-01 Go lab",
	}
	if err := event.Sign(secretKey); err != nil {
		panic(err)
	}

	var canonical any
	if err := json.Unmarshal(event.Serialize(), &canonical); err != nil {
		panic(err)
	}
	canonicalJSON, err := json.Marshal(canonical)
	if err != nil {
		panic(err)
	}
	eventJSON, err := json.MarshalIndent(event, "", "  ")
	if err != nil {
		panic(err)
	}

	fmt.Println("pubkey:", event.PubKey)
	fmt.Println("canonical preimage:", string(canonicalJSON))
	fmt.Println("event id:", event.ID)
	fmt.Println("event JSON:", string(eventJSON))
	fmt.Println("id verifies:", event.CheckID())
	signatureValid, err := event.CheckSignature()
	if err != nil {
		panic(err)
	}
	fmt.Println("signature verifies:", signatureValid)
}

File: examples/nostr-basics/01-event-envelope/rust/src/main.rs
use std::error::Error;

use nostr::{EventBuilder, JsonUtil, Keys, Kind, Tag};
use serde_json::json;

fn main() -> Result<(), Box<dyn Error>> {
    let keys = Keys::generate();
    let tag = Tag::parse(["t", "nip-01"])?;
    let event = EventBuilder::new(Kind::TextNote, "hello from the NIP-01 Rust lab")
        .tags([tag])
        .sign_with_keys(&keys)?;

    let canonical = json!([
        0,
        event.pubkey.to_hex(),
        event.created_at.as_secs(),
        event.kind.as_u16(),
        event.tags,
        event.content,
    ]);
    let canonical_json = serde_json::to_string(&canonical)?;

    println!("pubkey: {}", event.pubkey.to_hex());
    println!("canonical preimage: {canonical_json}");
    println!("event id: {}", event.id.to_hex());
    println!("event JSON: {}", event.as_json());
    println!("id verifies: {}", event.verify_id());
    println!("signature verifies: {}", event.verify_signature());

    Ok(())
}
